### PR TITLE
[IOPID-615] : Hub Spid Login - Token Validation

### PR DESCRIPTION
**Pull request by claudiodall**
#### List of Changes
Created a new middleware that handles token validation (JWT from hub spid login). The middleware validates the token provided with the hub-spid-login public key and hub-spid-login issuer. If the token is validated, we make a final call to the hub-spid-login introspection endpoint.

#### How Has This Been Tested?
- Utilized hub spid login to generate the JWT.
- Invoked the Ping function to verify the flow.

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: README.md
Comment (arcogabbo) [inline README.md:83]: I think some variable are missing in this table. for example, `PRIMARY_PRIVATE_KEY` and `PRIMARY_PUBLIC_KEY` seems required in the config but are not present in this table

File: Ping/handler.ts
Comment (arcogabbo) [inline Ping/handler.ts:39]: am I wrong or are we decoding multiple times the JWT in this middlewares? I suggest creating a middleware that decodes the JWT and puts it into `req.user` for example. this way we can forward the already validated JWT to other middlewares.

**NOTE**: this endpoint has to be removed before the final release. it is used only for demo purposes

EDIT 1: req.user is not available and passing a validated parameter to more middlewares requires another middleware combinator

File: clients/hubSpidLogin.ts
Comment (arcogabbo) [inline clients/hubSpidLogin.ts]: this can be removed
```suggestion
```

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (arcogabbo) [inline utils/middlewares/hsl-jwt-validation-middleware.ts]:
```suggestion
```

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (arcogabbo) [inline utils/middlewares/hsl-jwt-validation-middleware.ts]: instead of returning a Promise and using `tryCatch` after I suggest this refactor which returns a `TaskEither` container. also, in the previous implementation there was a map which put errors inside the right part.
```suggestion
type IjwtIntrospectionCall = (
  token: NonEmptyString,
  config: IConfig
) => TE.TaskEither<
  IResponseErrorInternal | IResponseErrorForbiddenNotAuthorized,
  IResponseSuccessJson<IntrospectSuccessResponse>
>;
export const introspectionCall: IjwtIntrospectionCall = (token, config) =>
  pipe(
    TE.tryCatch(
      () =>
        getHubSpidLoginClient(
          config.HUB_SPID_LOGIN_CLIENT_BASE_URL
        ).introspectToken({
          body: {
            token
          }
        }),
      flow(E.toError, () =>
        ResponseErrorInternal(
          `Something went wrong while calling the introspection endpoint`
        )
      )
    ),
    TE.chain(
      flow(
        TE.fromEither,
        TE.mapLeft(errors =>
          ResponseErrorInternal(readableReportSimplified(errors))
        ),
        TE.chainW(response => {
          switch (response.status) {
            case 200:
              return response.value.active
                ? TE.right(ResponseSuccessJson(response.value))
                : TE.left<
                    | IResponseErrorForbiddenNotAuthorized
                    | IResponseErrorInternal
                  >(ResponseErrorForbiddenNotAuthorized);
            case 403:
              return TE.left<
                IResponseErrorForbiddenNotAuthorized | IResponseErrorInternal
              >(ResponseErrorForbiddenNotAuthorized);
            default:
              return TE.left(ResponseErrorInternal(`Something went wrong`));
          }
        })
      )
    )
  );
```

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (arcogabbo) [inline utils/middlewares/hsl-jwt-validation-middleware.ts]: following the refactor above, we can simplify this pipeline
```suggestion
        introspectionCall(token, config),
        TE.fold(
          _ => TE.left(new Error("Something went wrong")),
          // active is always true if we are in this rail
          _ => TE.right(jwtDecoded as IHslJwtPayloadExtended)
        )
```

File: utils/middlewares/__tests__/hsl-jwt-validation.test.ts
Comment (arcogabbo) [inline utils/middlewares/__tests__/hsl-jwt-validation.test.ts]: I suggest to enrich this test. also the JWT provided in the config mock expires the 10th of August. another comment will address this problem
```suggestion
describe(`Hub Spid Login JWT Validation Middleware`, () => {
  it(`Should validate JWT and call hub-spid-login for introspection call,
    - Should return JWT payload 
    - Success -`, async () => {
    const mockIntrospectionCall = jest
      .spyOn(hslUtils, "introspectionCall")
      .mockReturnValueOnce(TE.right(ResponseSuccessJson({ active: true })));
    const token = config.HUB_SPID_LOGIN_MOCK_TOKEN;

    const jwtValidation = hslJwtValidation(token, config);
    const result = await jwtValidation(token)();

    expect(E.isRight(result)).toBe(true);
    expect(mockIntrospectionCall).toHaveBeenCalledTimes(1);
    expect(mockIntrospectionCall).toHaveBeenCalledWith(token, config);

    const resultPayload = E.fold(
      () => null,
      payload => payload
    )(result);

    expect(resultPayload).toMatchObject({
      name: "Carla",
      family_name: "Rossi",
      fiscal_number: "ISPXNB32R82Y766D"
    });
  });
});
```

File: __mocks__/config.mock.ts
Comment (arcogabbo) [inline __mocks__/config.mock.ts:1]: I would like to suggest this implementation to have more control on the values. the JWT would also be generated dynamically, preventing the test to fail, for example, for expire timestamps.
```typescript
import { NonEmptyString } from "@pagopa/ts-commons/lib/strings";
import { sign } from "jsonwebtoken";
import { HSLConfig, IConfig, JWTConfig } from "../utils/config";
import { FeatureFlagEnum } from "../utils/featureFlags/featureFlags";
import { generateKeyPairSync } from "crypto";

const aFiscalCode = "ISPXNB32R82Y766D";
const aName = "Carla";
const aFamilyName = "Rossi";
const iconfig = {
  AzureWebJobsStorage: "azureWebJobsStorage" as NonEmptyString,
  BETA_TESTERS: [aFiscalCode],
  COSMOSDB_KEY: "cosmosdbKey" as NonEmptyString,
  COSMOSDB_NAME: "cosmosdbName" as NonEmptyString,
  COSMOSDB_URI: "cosmosdbUri" as NonEmptyString,
  FF_API_ENABLED: FeatureFlagEnum.ALL,
  QueueStorageConnection: "queueStorageConnection" as NonEmptyString,
  isProduction: false
};

const {
  privateKey: hslPrivateKey,
  publicKey: hslPublicKey
} = generateKeyPairSync("rsa", {
  modulusLength: 2048,
  publicKeyEncoding: {
    type: "spki",
    format: "pem"
  },
  privateKeyEncoding: {
    type: "pkcs8",
    format: "pem"
  }
});

const hslIssuer = "SPID";

const jwt = sign(
  { name: aName, family_name: aFamilyName, fiscal_number: aFiscalCode },
  hslPrivateKey,
  { algorithm: "RS256", issuer: hslIssuer }
);

export const jwtConfig: JWTConfig = {
  BEARER_AUTH_HEADER: "authorization" as NonEmptyString,
  HUB_SPID_LOGIN_JWT_ISSUER: hslIssuer as NonEmptyString,
  HUB_SPID_LOGIN_JWT_PUB_KEY: hslPublicKey as NonEmptyString,
  ISSUER: "PAGOPA" as NonEmptyString,
  JWT_TTL: 900,
  PRIMARY_PRIVATE_KEY: "primaryPrivateKey" as NonEmptyString,
  PRIMARY_PUBLIC_KEY: "primaryPublicKey" as NonEmptyString
};

export const hslConfig: HSLConfig = {
  HUB_SPID_LOGIN_MOCK_TOKEN: jwt as NonEmptyString,
  HUB_SPID_LOGIN_API_KEY: "hslApiKey" as NonEmptyString,
  HUB_SPID_LOGIN_CLIENT_BASE_URL: "http://localhost:9090" as NonEmptyString
};

export const config: IConfig = {
  ...iconfig,
  ...jwtConfig,
  ...hslConfig
};
```

**Review by arcogabbo — CHANGES_REQUESTED** [review]
Very nice work, I only have some suggestions

File: __mocks__/config.mock.ts
Comment (claudiodall) [reply __mocks__/config.mock.ts:1]: Ok, thanks

File: README.md
Comment (claudiodall) [reply README.md:83]: I've removed those variables since they are leftovers from the previous task, where we were configuring all the variables to generate the token for the magic link.

**Review by arcogabbo — APPROVED** [review]
LGTM. Let's wait for @BurnedMarshal approval before merging.

File: Ping/handler.ts
Comment (BurnedMarshal) [inline Ping/handler.ts]:
```suggestion
```

File: .env.example
Comment (BurnedMarshal) [inline .env.example]: Are some of this CF real?

File: package.json
Comment (BurnedMarshal) [inline package.json]: This command generate even the client not only the definitions, we should rename the command accordingly.
```suggestion
    "generate:hub-spid-login": "rimraf ./generated/definitions/hub-spid-login && shx mkdir -p ./generated/definitions/hub-spid-login && gen-api-models --api-spec ./openapi/hub-spid-login.yaml --no-strict --out-dir ./generated/definitions/hub-spid-login --request-types --response-decoders --client",
```

File: utils/config.ts
Comment (BurnedMarshal) [inline utils/config.ts]: This parameter shouldn't be required. In production must be disabled.
We can use this declaration
```typescript
export type HSLConfig = t.TypeOf<typeof HSLConfig>;
export const HSLConfig = t.intersection([
  t.type({
    HUB_SPID_LOGIN_API_KEY: NonEmptyString,
    HUB_SPID_LOGIN_CLIENT_BASE_URL: NonEmptyString
  }),
  t.partial({
    HUB_SPID_LOGIN_MOCK_TOKEN: NonEmptyString
  })
]);
```

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (BurnedMarshal) [inline utils/middlewares/hsl-jwt-validation-middleware.ts:42]: We should improve the client using an timeout abortable fetch.
You can see the implementation in [here](https://github.com/pagopa/io-backend/blob/5c594aeff63db9294ecf9a084655efcd2f5791b3/src/config.ts#L413)
The Application gateway has a timeout of 10 seconds, if the request take more than timeout time we can safely release the socket freeing resources.

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (BurnedMarshal) [inline utils/middlewares/hsl-jwt-validation-middleware.ts:107]: Why? If the token will be invalidated `active=false`

**Review by BurnedMarshal — CHANGES_REQUESTED** [review]
Thanks for the PR, only some minor consideration.

File: Ping/handler.ts
Comment (BurnedMarshal) [reply Ping/handler.ts:39]: The 2 operation seems to be different. The first only decode the JWT reading the payload instead the second verify the JWT signature and call the HSL introspection endpoint. Only the second operation has a high operating cost.

File: utils/config.ts
Comment (claudiodall) [reply utils/config.ts]: Ok, thanks

File: .env.example
Comment (claudiodall) [reply .env.example]: Let me remove them

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (claudiodall) [reply utils/middlewares/hsl-jwt-validation-middleware.ts:107]: If the token after introspection is not active, it returns 200 and active = false. Do you mean to change the error message?

Comment (BurnedMarshal) [reply utils/middlewares/hsl-jwt-validation-middleware.ts:107]: The error message it's ok. But if the value of `active` is `false` the handler should return `unauthorized`. This implementation seems to allow invalidated tokens. Can we add a test for this scenario please?

Comment (arcogabbo) [reply utils/middlewares/hsl-jwt-validation-middleware.ts:107]: based on the [current implementation](https://github.com/pagopa/hub-spid-login-ms/blob/7e029ea0d8d27bf80fe1b1214974c3048bb9674f/src/handlers/token.ts#L124) of the introspect endpoint we always receive `active: true` when the response has status code of 200. however if we want to return unauthorized on this middleware we can filter the `IResponse` returned by `introspectionCall` function.

Comment (BurnedMarshal) [reply utils/middlewares/hsl-jwt-validation-middleware.ts:107]: My bad, if `active=false` with status code `200` the response is `ResponseErrorForbiddenNotAuthorized` ([here](https://github.com/pagopa/io-web-profile-backend/pull/8/files#diff-982b6d6010152452987d93baa8e20b94c0133100dd7c3c2c0cf7df52aae46679R63-R68)). But the implementation seems to return always `true` with that http status code. So the type for the generated client is wrong in [here](https://github.com/pagopa/io-web-profile-backend/pull/8/files#diff-e13b8fe59f2873727f8133419563d711b0b3d031ef9af3b7b529e3557aaafe39R48)
Maybe should be like
```yaml
IntrospectSuccessResponse:
  type: object
  properties:
    active:
      type: boolean
      enum:
        - true
  required:
    - active
  example:
    active: true
IntrospectForbiddenResponse:
  type: object
  properties:
    active:
      type: boolean
      enum:
        - false
  required:
    - active
  example:
    active: false
```

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (claudiodall) [reply utils/middlewares/hsl-jwt-validation-middleware.ts:42]: Ok, sure, thanks

File: utils/middlewares/hsl-jwt-validation-middleware.ts
Comment (claudiodall) [reply utils/middlewares/hsl-jwt-validation-middleware.ts:107]: Okay, fixed as requested and added the test.

File: openapi/hub-spid-login.yaml
Comment (arcogabbo) [inline openapi/hub-spid-login.yaml]: this two lines have to be indented to the right by two spaces

Comment (claudiodall) [reply openapi/hub-spid-login.yaml]: Fixxed

File: openapi/hub-spid-login.yaml
Comment (BurnedMarshal) [inline openapi/hub-spid-login.yaml]:
```suggestion
            - false

```

**Review by BurnedMarshal — APPROVED** [review]
LGTM

**Pull request merged**
